### PR TITLE
Debugger File Selection Improvements

### PR DESCRIPTION
**Pull request by EBatTiVo**
Load files using canonical names. (Resolve symlinks so that IDEA does not open
a symlink as a duplicate buffer.)
Use implicit class path entries when searching for relative paths. (haxe/std)

### CODE REVIEW

**Comment by EBatTiVo** [comment]
@sganapavarapu1 Please review.

**Comment by sganapavarapu1** [comment]
CI builds are failing

**Comment by EBatTiVo** [comment]
The CI builds are failing on 14.1.1 because the JetBrains site isn't serving the IDEA release for 14.1.1 at the moment.  It has nothing to do with the sources.  Please continue the review.  Thanks.

File: common/src/com/intellij/plugins/haxe/config/HaxeTarget.java [hunk 1]
```diff
@@ -107,4 +108,25 @@ public static void initCombo(@NotNull DefaultComboBoxModel comboBoxModel) {
...
+   */
+  @Nullable
+  public static HaxeTarget matchOutputTarget(String compilerTargetArgument) {
+    for (HaxeTarget t : HaxeTarget.values()) {
+      if (t.getCompilerFlag().equals(compilerTargetArgument)) {
+        return t;
+      }
+    }
+    // as3 is an old case.
+    if ("-as3".equals(compilerTargetArgument)) {
```
Comment (sganapavarapu1) [inline common/src/com/intellij/plugins/haxe/config/HaxeTarget.java]: I thought for '-as3', target should be FLASH

File: src/13.1/com/intellij/plugins/haxe/runner/debugger/HaxeDebugRunner.java [hunk 2]
```diff
@@ -804,18 +809,35 @@ public StackFrame(Project project, Module module,
...
+          // an appropriate file.
+          file = HaxelibClasspathUtils.findFileOnClasspath(module, mFileName);
+        } else if (matches.size() == 1) {
+          // Got one.  If it's a good file, keep it.  Otherwise, try to pick it
+          // out of the classpath.
+          VirtualFile possible = matches.iterator().next();
+          file = possible.isValid() ? possible : HaxelibClasspathUtils.findFileOnClasspath(module, possible.toString());
+        } else {
+          // Too many matches. Get the first that occurs on the classpath.
+          file = HaxelibClasspathUtils.findFirstFileOnClasspath(module, matches);
```
Comment (sganapavarapu1) [inline src/13.1/com/intellij/plugins/haxe/runner/debugger/HaxeDebugRunner.java]: is this classPath same as the classPath that the compiler uses?
if it is, then picking the 1st one in classPath makes perfect sense
the reason I asked that question is, I wasn't sure if HaxelibClasspathUtils returns the classPath that the application being debugged is compiled/run with... I could dig around, but asking you as you're already familiar with that code while implementing this

File: common/src/com/intellij/plugins/haxe/config/HaxeTarget.java (hunk 1 above)
Comment (EBatTiVo) [inline common/src/com/intellij/plugins/haxe/config/HaxeTarget.java]: Really?  AS3 is technically a superset of JavaScript.  Does it compile to a .swf file?  I could go either way, really.

File: src/13.1/com/intellij/plugins/haxe/runner/debugger/HaxeDebugRunner.java [hunk 3]
```diff
@@ -804,18 +809,35 @@ public StackFrame(Project project, Module module,
...
+        }
+
+        // Now, work around the fact that IDEA treats symlinks as separate files.
+        // XXX: This should be controlled via an UI option.
+        if (null != file && file.is(VFileProperty.SYMLINK)) {
```
Comment (sganapavarapu1) [inline src/13.1/com/intellij/plugins/haxe/runner/debugger/HaxeDebugRunner.java]: nice, defensive strategy
just curious: how do we control this via an UI option? can't think something like that's possible at present - are you proposing a UI enhancement story?

File: src/13.1/com/intellij/plugins/haxe/runner/debugger/HaxeDebugRunner.java (hunk 2 above)
Comment (sganapavarapu1) [inline src/13.1/com/intellij/plugins/haxe/runner/debugger/HaxeDebugRunner.java]: ah! I jumped ahead - I see your implementation in HaxelibClasspathUtils... ignore my above question

File: src/13.1/com/intellij/plugins/haxe/runner/debugger/HaxeDebugRunner.java (hunk 3 above)
Comment (EBatTiVo) [inline src/13.1/com/intellij/plugins/haxe/runner/debugger/HaxeDebugRunner.java]: Yes, I would like to add a few checkboxes to the openfl/debugger pane.  Right now, I'm just putting comments in the code.

File: common/src/com/intellij/plugins/haxe/config/HaxeTarget.java (hunk 1 above)
Comment (sganapavarapu1) [inline common/src/com/intellij/plugins/haxe/config/HaxeTarget.java]: in general, I've seem that our AS3 code (in Bravo & pre-Haxe Encore) built by IDEA producing either .swf or .swc file (depending on whether it is an application or a library). 
I'm not 100% (as I didn't test it) whether same applies when compiling from Haxe to AS3 target - though I'd assume it to be generally true ... that's what other developers usually mention when working with IntelliJ and compiling to AS3 target

Comment (sganapavarapu1) [inline common/src/com/intellij/plugins/haxe/config/HaxeTarget.java]: let's mark it FLASH

**Comment by sganapavarapu1** [comment]
I approve this for merge (assuming you'd change that AS3/JAVA_SCRIPT to AS3/FLASH)

File: common/src/com/intellij/plugins/haxe/config/HaxeTarget.java (hunk 1 above)
Comment (EBatTiVo) [inline common/src/com/intellij/plugins/haxe/config/HaxeTarget.java]: OK.  I'll change it.  But first, I'll check the lime tool to see what it does with -as3.

**Comment by EBatTiVo** [comment]
Thanks.

File: .travis.yml [hunk 4]
```diff
@@ -15,6 +15,6 @@ script: make test
 env:
     - IDEA_VERSION=13.1.6
     - IDEA_VERSION=14.0.4
-    - IDEA_VERSION=14.1.1
+    - IDEA_VERSION=14.1.4
```
Comment (sganapavarapu1) [inline .travis.yml]: I was occupied with a P1/S1 bug, and tried to review after travis builds pass - assumed that some unit test is failing
sorry that I didn't pay attention to why it was failing

File: common/src/com/intellij/plugins/haxe/config/HaxeTarget.java (hunk 1 above)
Comment (sganapavarapu1) [inline common/src/com/intellij/plugins/haxe/config/HaxeTarget.java]: Sure,

**Pull request merged**